lib/src/component/with_transition.dart

```diff
@@ -0,0 +1,410 @@
+  /// transition states that are confined within the [WithTransition] component without having to
+  /// utilize state change callbacks to update your component's state.
+  ///
+  /// __NOTE:__ These values will override any prop values set on the child itself if the
+  /// key(s) in the `Map` are the same.
+  ///
+  /// __Example:__
+  ///
+  /// ```dart
+  /// UiFactory<WithTransitionExampleProps> WithTransitionExample = uiFunction(
+  ///   (props) {
+  ///     final isShown = useState(false);
+  ///
+  ///     (WithTransition()
+  ///       ..isShown = isShown.value
+  ///       ..childPropsByPhase = useMemo(() => {
+  ///         TransitionPhase.PRE_SHOWING: domProps()..className = 'before-showing',
+  ///         TransitionPhase.SHOWING: domProps()..className = 'showing',
+  ///         TransitionPhase.SHOWN: domProps()..className = 'shown',
+  ///         TransitionPhase.HIDING: domProps()..className = 'hiding',
+  ///         TransitionPhase.HIDDEN: domProps()..className = 'hidden',
+  ///       })
+  ///     )(
+  ///       // The child that has CSS transitions
+  ///     )
+  ///   },
+  ///   $WithTransitionExampleConfig, // ignore: undefined_identifier
+  /// );
+  /// ```
+  Map<TransitionPhase, Map> childPropsByPhase;
+
+  /// The amount of time to wait for a CSS transition to complete before assuming something went wrong
+  /// and canceling the subscription.
+  ///
+  /// If the CSS transition duration on the node is longer than `1s`, you will need to set this to a Duration
+  /// that is greater than or equal to the expected CSS duration of the node.
+  ///
+  /// > Default `const Duration(seconds: 1)`
+  Duration transitionTimeout;
+}
+
+class WithTransitionProps = UiProps with v2.TransitionPropsMixin, WithTransitionPropsMixin;
+
+/// A wrapper component that hooks into CSS transition(s) present on a single child passed to it.
+///
+/// For example, say you have an element that "fades" in/out using the `fade` CSS class to apply
+/// a CSS `transition`, and an `in` CSS class that - when present - changes the opacity from `0` to `1`.
+///
+/// Using this component, you can simply toggle the value of [WithTransitionPropsMixin.isShown] from `false` to
+/// `true` using your own state to make the node "fade in", and hook into the available prop callbacks like
+/// `onDidShow`, `onDidHide` to know when the CSS transition has completed.
+///
+/// ```dart
+/// import 'package:over_react/over_react.dart';
+/// import 'package:over_react/components.dart' show WithTransition;
+///
+/// mixin WithTransitionExampleProps on UiProps {
+///   bool initiallyShown;
+/// }
+///
+/// UiFactory<WithTransitionExampleProps> WithTransitionExample = uiFunction(
+///   (_props) {
+///     final defaultProps = useMemo(() {
+///       return WithTransitionExample()..initiallyShown = false;
+///     }, const []);
+///     final props = defaultProps..addProps(_props);
+///
+///     final isShown = useState(props.initiallyShown);
```

Comment:
#nit the preferred way to default props should be via null-awares (for a number of reasons, including promoting the values to non-nullable once nnbd is enabled)

```suggestion
///     final initiallyShown = props.initiallyShown ?? false;
///
///     final isShown = useState(initiallyShown);
```

Reply:
Addressed in https://github.com/Workiva/over_react/pull/621/commits/967d93dd4c6cbb6fac367b02d128a6002613d8b8